---
title: "Project Summary"
output:
  html_document:
    toc: true
    theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup

### Loading required libraries
First, we'll load all required libraries

```{r libraries, results='hide', message=FALSE, warning=FALSE}
library(HGNChelper)
library(openxlsx)
library(expss)
library(ggpubr)
library(dplyr)
library(scales)

```

```{r universeskip,include=FALSE}
load("output/Data/universe_df.rda")
```

### Initialising data frame of all protein-coding genes

Next, we'll initialise our data frame of all protein-coding genes, universe_df, which we'll be using throughout the project. 

To do this, we'll load in HGNC's spreadsheet of protein-coding genes, downloaded from [their website](https://www.genenames.org/cgi-bin/statistics), updating gene symbols using, and taking only the columns we want. 

```{r universe_init, eval = FALSE}
#read protein-coding gene list
hgnc<- read.csv("Gene_lists/Universe/gene_with_protein_product.txt",sep = "\t", comment.char = "#",stringsAsFactors = FALSE)

#take only needed columns
hgnc$symbol<- checkGeneSymbols(hgnc$symbol)[[3]]
universe_df <- data.frame(hgnc$hgnc_id,hgnc$symbol,hgnc$name,hgnc$gene_family,hgnc$location,hgnc$mgd_id)
names(universe_df) <- c("hgnc_id","gene","gene_name","gene_family","location","mgd_id")
rm(hgnc)

```

## Creating an easy-to-mine spreadsheet collating data on human phenotype, mouse and cell KO, ExAC population genetics data etc.

### Getting disease genes from OMIM

A complete list of OMIM genes was downloaded from [OMIM Data Downloads page](https://www.omim.org/downloads/) with license on 2018-03-10 (file: genemap2.txt and mim2gene.txt). This list of genes was filtered to create a list of Mendelian disease genes with clinically relevant phenotype.

First, we load in data in genemap2:

```{r omim_df_parse, eval=FALSE}
# Load in genemap2
omim <- read.csv(file = "Gene_lists/OMIM/genemap2.txt", sep = "\t", comment.char = "#",stringsAsFactors = FALSE)
#take only needed columns
omim <- data.frame(omim$Mim.Number,omim$Approved.Symbol,omim$Gene.Name,omim$Phenotypes,omim$Mouse.Gene.Symbol.ID)
omim <- setNames(omim, c("Mim.Number","Approved.Symbol","Gene.Name","Phenotypes","Mouse.Gene.Symbol.ID"))
#changing types to character
omim$Phenotypes <- levels(omim$Phenotypes)[omim$Phenotypes]
omim$Approved.Symbol <- levels(omim$Approved.Symbol)[omim$Approved.Symbol]
omim$Mouse.Gene.Symbol.ID <- levels(omim$Mouse.Gene.Symbol.ID)[omim$Mouse.Gene.Symbol.ID]
```

Next, we're going to filter these genes so we only include genes which meet our criteria for inclusion in ‘clinically relevant mendelian disease genes’ OMIM list:

* Annotated with phenotype mapping key (3) (“molecular basis is known”) 
* Phenotype does not contain code for ‘nondisease’ [],’susceptibility’ [2],’provisional link’ ?, somatic mutation (text)

[see OMIM FAQ]("https://www.omim.org/help/faq#1_6")

```{r omim_df_filter1, eval= FALSE}

# keep only rows containing phenotype mapping key (3) "molecular basis is known"
omim <- omim[which(grepl(pattern="\\(3)",x=omim$Phenotypes,ignore.case=FALSE)==TRUE),]

#look through phenotypes, filter out any nondiseases, susceptibilities, provisional links, somatic mutations
for (j in 1:length(omim$Phenotypes)) {
  omim$Phenotypes[j] <- strsplit(toString(omim$Phenotypes[j]),"; ")
}
for (i in 1:length(omim$Phenotypes)) {
  omim$Phenotypes[[i]]<- omim$Phenotypes[[i]][which(!grepl(pattern="\\[",x=omim$Phenotypes[[i]],ignore.case=FALSE)&!grepl(pattern="\\{",x=omim$Phenotypes[[i]],ignore.case=FALSE)&!grepl(pattern="\\?",x=omim$Phenotypes[[i]],ignore.case=FALSE)&!grepl(pattern="somatic",x=omim$Phenotypes[[i]],ignore.case=FALSE)&!grepl(pattern="Somatic",x=omim$Phenotypes[[i]],ignore.case=FALSE)&grepl(pattern="\\(3)",x=omim$Phenotypes[[i]],ignore.case=FALSE))]
}
omim <- omim[-which(lengths(omim$Phenotypes)==0),]
rm(i,j)
```

Now we

* update gene symbols with HGNChelper
* 2 genes have 2 entries in mim2gene so we concatenate phenotypes associated with these genes
* exclude non protein-coding genes by checking OMIM genes are in our universe gene list


```{r omim_df_filter2, eval=FALSE}
#updating gene names
omim$Gene <- checkGeneSymbols(omim$Approved.Symbol,unmapped.as.na=TRUE)[[3]]

# manually fixing a few HGNChelper didn't catch
omim$Gene[which(omim$Approved.Symbol=="Sep-12")]<- "SEPT12"
omim$Gene[which(omim$Approved.Symbol=="Sep-09")]<- "SEPT9"
omim$Gene[which(omim$Approved.Symbol=="NECTIN 4")] <- "NECTIN4"
omim$Gene[which(omim$Approved.Symbol=="NTF4 ")] <- "NTF4"
#remaining unmapped IDS are not protein-coding genes (disease-associated loci, non-coding genes etc) so we exclude them
omim <- omim[-which(is.na(omim$Gene)),]

#checking for duplicate genes, concatenating rows with duplicates, deleting duplicates
dups <- which(duplicated(omim$Gene)==TRUE)
for (p in 1:length(dups)) {
  genedups <- omim$Gene[dups[p]]
  index <- which(omim$Gene==genedups)
  omim$Mim.Number[index[1]]<- paste(omim$Mim.Number[index[1]],omim$Mim.Number[index[2]],sep=", ")
  omim$Phenotypes[index[1]]<- paste(omim$Phenotypes[index[1]],omim$Phenotypes[index[2]],sep="; ")
  if (length(omim$Mouse.Gene.Symbol.ID[index[which(nchar(omim$Mouse.Gene.Symbol.ID[index])>0)]])>0) {
    omim$Mouse.Gene.Symbol.ID[index[1]]<- omim$Mouse.Gene.Symbol.ID[index[which(nchar(omim$Mouse.Gene.Symbol.ID[index])>0)]]
  }
}
if (length(dups)>0){
  omim <- omim[-dups,]
}
rm(dups,genedups,index,p)
#removing non protein-coding genes (noncoding RNAs, complex loci, immunoglobulins etc)
omim <- omim[-which(!omim$Gene%in%universe_df$gene&!grepl(pattern="///",omim$Gene)),]
```

Now we annotate these genes with the inheritance pattern of any associated phenotypes

```{r omim_df_inheritance, eval=FALSE}
#getting inheritances
# Extend OMIM data
omim$MT <- ifelse(omim$Phenotypes == "",NA,ifelse(grepl(pattern = ", Mitochondrial", x = omim$Phenotypes, ignore.case = FALSE),"Y","N"))
omim$AR <- ifelse(omim$Phenotypes == "",NA,ifelse(grepl(pattern = "Autosomal recessive", x = omim$Phenotypes, ignore.case = FALSE),"Y","N"))
omim$AD <- ifelse(omim$Phenotypes == "", NA,ifelse(grepl(pattern = "Autosomal dominant", x = omim$Phenotypes, ignore.case = FALSE),"Y","N"))
omim$XLd <- ifelse(omim$Phenotypes == "",NA,ifelse(grepl(pattern = "X-linked dominant", x = omim$Phenotypes, ignore.case = FALSE),"Y","N"))
omim$XLr <- ifelse(omim$Phenotypes == "",NA,ifelse(grepl(pattern = "X-linked recessive", x = omim$Phenotypes, ignore.case = FALSE),"Y","N"))
omim$IS <- ifelse(omim$Phenotypes == "",NA,ifelse(grepl(pattern = "isolated cases", x = omim$Phenotypes, ignore.case = FALSE),"Y","N"))

#creating inheritance pattern column
omim$Inheritance_pattern <- rep(NA,length(omim$Gene))
omim$Inheritance_pattern[which(omim$MT=="Y")] <- "MT"
omim$Inheritance_pattern[which(omim$XLd == "Y")]<- ifelse(is.na(omim$Inheritance_pattern[which(omim$XLd == "Y")]),"XLd",paste(omim$Inheritance_pattern[which(omim$XLd == "Y")],"XLd",sep=","))
omim$Inheritance_pattern[which(omim$XLr == "Y")]<- ifelse(is.na(omim$Inheritance_pattern[which(omim$XLr == "Y")]),"XLr",paste(omim$Inheritance_pattern[which(omim$XLr == "Y")],"XLr",sep=","))
omim$Inheritance_pattern[which(omim$AR == "Y")]<- ifelse(is.na(omim$Inheritance_pattern[which(omim$AR == "Y")]),"AR",paste(omim$Inheritance_pattern[which(omim$AR == "Y")],"AR",sep=","))
omim$Inheritance_pattern[which(omim$AD == "Y")]<- ifelse(is.na(omim$Inheritance_pattern[which(omim$AD == "Y")]),"AD",paste(omim$Inheritance_pattern[which(omim$AD == "Y")],"AD",sep=","))
omim$Inheritance_pattern[which(omim$IS == "Y")]<- ifelse(is.na(omim$Inheritance_pattern[which(omim$IS == "Y")]),"IS",paste(omim$Inheritance_pattern[which(omim$IS == "Y")],"IS",sep=","))
```

And we can save our filtered OMIM gene list for easy access/use later

```{r omim_save, eval=FALSE}
#taking only needed rows, saving to spreadsheet
omim <- omim[,c(1,4,5,6,13)]
write.xlsx(omim, file= "output/spreadsheets/omim_filtered_with_inheritances.xlsx", row.names = FALSE,col.names = TRUE)

```

### get exac scores for protein coding genes

Next we'll load in ExAC gene constraint data 

```{r exac1,eval=FALSE}
exac <- read.table("Gene_lists/ExAC/fordist_cleaned_exac_r03_march16_z_pli_rec_null_data.txt",header = TRUE,fill = TRUE)
exac <- exac[,c(2,17,18,20)]
```

Some ExAC gene symbols map to the same gene symbol when updated, here we remove the duplicated genes, as well as removing any ExAC genes that aren't protein coding

```{r exac2, eval=FALSE }

genecheck <- checkGeneSymbols(exac$gene,unmapped.as.na=FALSE)
baddups <- genecheck[which(duplicated(genecheck[[3]])==TRUE),3]
a<- genecheck[which(genecheck[[3]]%in%baddups),]
b<- which(genecheck[[1]]%in%a[which(a[[2]]==FALSE),1])
genecheck <- genecheck[-b,]
exac <-exac[-b,]
exac$gene <- genecheck[[3]]
rm(a,b,baddups,genecheck)

#removing exac genes that aren't protein coding
exac <- exac[-which(!exac$gene%in%universe_df$gene&!grepl(pattern="///",exac$gene)),]

```


### regional constraint

Next we'll load data from three different papers defining genes with regional constraint

* A map of constrained coding regions in the human genome.
James M Havrilla, Brent S Pedersen, Ryan M Layer, Aaron R Quinlan
bioRxiv 220814; doi: https://doi.org/10.1101/220814

[github](https://github.com/quinlan-lab/ccrhtml)

I used the CCR Bed File from the github repo to put together details on the highest CCR contained in each protein-coding genes in universe, as well as which genes had a CCR>99.

```{r ccr, eval=FALSE}
#havrilla CCR
ccr <- as.data.frame(read.table("Gene_lists/Regional_Constraint/ccrs.autosomes.v2.20180420.bed",header = FALSE, sep="\t",stringsAsFactors=FALSE))
ccr95 <- ccr[which(ccr$V4>95),]
ccr95$V5<-checkGeneSymbols(ccr95$V5,unmapped.as.na=FALSE,map=hgnc.table)[[3]]
save(ccr95, file="output/Data/ccr95.rda", compress="bzip2")
ccr99 <- ccr[which(ccr$V4>99),]
ccr99$V5<-checkGeneSymbols(ccr99$V5,unmapped.as.na=FALSE,map=hgnc.table)[[3]]
save(ccr99, file="output/Data/ccr99.rda", compress="bzip2")

mult99<-read.table("Gene_lists/Regional_Constraint/genesw99CCR(supp_table_1).tsv", sep = '\t', header = TRUE,stringsAsFactors=FALSE)
ccr99_genes <- data.frame(gene=unique(ccr99$V5))
ccr99_genes$n <- ifelse(ccr99_genes$gene%in%mult99$gene,vlookup(ccr99_genes$gene,mult99,result_column="n",lookup_column = "gene"),1)
save(ccr99_genes, file="output/Data/ccr99_genes.rda", compress="bzip2")
rm(mult99)

highest_ccr <- data.frame(gene=unique(ccr$V5))
highest_ccr$ccr <- rep(NA,length(highest_ccr$gene))
highest_ccr$ccr <- unlist(lapply(highest_ccr$gene,function(x) max(ccr$V4[which(ccr$V5==x)])))
highest_ccr$gene<-checkGeneSymbols(highest_ccr$gene,unmapped.as.na=FALSE,map=hgnc.table)[[3]]
save(highest_ccr, file="output/Data/highest_ccr.rda", compress="bzip2")
```

Next we load in data from 

* Regional missense constraint improves variant deleteriousness prediction
Kaitlin E. Samocha, Jack A. Kosmicki, Konrad J. Karczewski, Anne H. O'Donnell-Luria, Emma Pierce-Hoffman, Daniel G. MacArthur, Benjamin M. Neale, Mark J. Daly
bioRxiv 148353; doi: https://doi.org/10.1101/148353

Here I load in missense constraint data using Table S4 downloaded from [here](https://www.biorxiv.org/content/early/2017/06/12/148353.figures-only)

I annotate each protein-coding gene with the lowest gamma contained in the gene as defined by Samocha et al.

```{r samocha, eval=FALSE }
#samocha regional missense constraint
samocha<-read.xlsx("Gene_lists/Regional_Constraint/148353-3.xlsx",sheet=2)
genes_w_var <- data.frame(gene=unique(samocha$gene))
genes_w_var$lowest_gamma <- unlist(lapply(genes_w_var$gene,function(x) min(samocha$obs_exp[which(samocha$gene==x)])))
genes_w_var$gene<-checkGeneSymbols(genes_w_var$gene,unmapped.as.na=FALSE,map=hgnc.table)[[3]]
rm(samocha)
```

Lastly, we'll load in data from 

* Zeynep Coban-Akdemir, Janson J. White, Xiaofei Song, Shalini N. Jhangiani, Jawid M. Fatih, Tomasz Gambin, Yavuz Bayram, Ivan K. Chinn, Ender Karaca, Jaya Punetha, Cecilia Poli, Eric Boerwinkle, Chad A. Shaw, Jordan S. Orange, Richard A. Gibbs, Tuuli Lappalainen, James R. Lupski, Claudia M.B. Carvalho,
Identifying Genes Whose Mutant Transcripts Cause Dominant Disease Traits by Potential Gain-of-Function Alleles,
The American Journal of Human Genetics,Volume 103, Issue 2; doi: https://doi.org/10.1016/j.ajhg.2018.06.009.


We load in Table S4 from the paper, a list of 1,996 protein coding genes they defined as having NMD-escape constraint, as well as their rank

```{r nmd, eval=FALSE}
#coban-akdemir NMD- constraint
nmd_min <-read.xlsx("Gene_lists/Regional_Constraint/1-s2.0-S0002929718302039-mmc5.xlsx",startRow=2)
nmd_min<-nmd_min[,c(1,2)]
nmd_min$gene<-checkGeneSymbols(nmd_min$gene,unmapped.as.na=FALSE,map=hgnc.table)[[3]]
```


### Mouse KO phenotypes

Next we're going to load in data on mouse KO phenotypes, and find which protein coding genes are associated with lethality in recessive KO mice. We use two data sources:

1. Mouse Genome Informatics (MGI)
2. International Mouse Phenotyping Consortium (IMPC)


#### MGI

All spreadsheets downloaded from [MGI Data and Statistical Reports page](http://www.informatics.jax.org/downloads/reports/index.html)

MGI_PhenoGenoMP contains all Mammalian Phenotype (MP) terms associated with every allele in MGI. We load in that spreadsheet, and remove any alleles which involve more than one protein-coding gene

```{r mgi_parse,eval=FALSE}
#load in all genes and phenotypes
mgi_df <- read.xlsx("Gene_lists/MGI_MPs/MGI_PhenoGenoMP.xlsx")
mgi_df <- mgi_df[,c(1,2,6,4)]
mgi_df<- mgi_df[which(!grepl(pattern = "\\,", x = mgi_df$MGI_ID, ignore.case = FALSE)),]
```

Next we check whether alleles are functional knockouts using the allele information spreadsheet

```{r mgi_parse2,eval=FALSE}
allele <- read.xlsx("Gene_lists/MGI_MPs/MGI_PhenotypicAllele.xlsx")
mgi_df$allele_info <- vlookup(mgi_df$`Allele.Symbol(s)`,allele,result_column="Allele.Attribute",lookup_column="Allele.Symbol")
mgi_df <- mgi_df[which(grepl(pattern="Null/knockout",x=mgi_df$allele_info,ignore.case=FALSE)|grepl(pattern="Hypomorph",x=mgi_df$allele_info,ignore.case=TRUE)),]
```

We use a manually curated list of lethal MP terms to check which protein-coding genes are associated with lethality in KO mice.

We keep all MP annotations in one column, and create another column that's just any lethal MP terms associated with each gene.

We also add a column with 'high level phens' to add another level of possible analysis later.

We then annotate MGI with gene symbols (mouse and human)

```{r mgi_parse3,eval=FALSE}
mgi <- data.frame(MGI_ID= unique(mgi_df$MGI_ID))

#which genes have a lethal phenotype
mgi_lethalphens <- read.xlsx("Gene_lists/MGI_MPs/MGI_lethal_phenotypes.xlsx")
mgi_df$is_lethal <- ifelse(mgi_df$MP_ID%in%mgi_lethalphens$MP.id,"Y","N")

#getting all other phens
mgi_allphens <- read.xlsx("Gene_lists/MGI_MPs/VOC_MammalianPhenotype.xlsx",startRow=1,colNames=FALSE)
names(mgi_allphens)<- c("MP.id","phenotype","definition")

#high level phens
mgi_highphens <- read.xlsx("Gene_lists/MGI_MPs/HMD_HumanPhenotype.xlsx")
mgi_highphens <- mgi_highphens[,c(6,7)]
names(mgi_highphens) <- c("MGI.id","phenotype")
#mp ids
mgi$all_MP_ID <- lapply(mgi$MGI_ID, function(x) unique(mgi_df$MP_ID[which(mgi_df$MGI_ID==x)]))
mgi$lethal_MP_ID <- lapply(mgi$MGI_ID, function(x) unique(mgi_df$MP_ID[which(mgi_df$MGI_ID==x&mgi_df$is_lethal=="Y")]))
mgi$all_MP_phen <- lapply(mgi$all_MP_ID, function(x) mgi_allphens$phenotype[which(mgi_allphens$MP.id%in%x)])
mgi$lethal_MP_phen <- lapply(mgi$lethal_MP_ID, function(x) mgi_lethalphens$phenotype[which(mgi_lethalphens$MP.id%in%x)])

mgi$high_MP_ID <- strsplit(vlookup(mgi$MGI_ID,mgi_highphens,result_column="phenotype",lookup_column="MGI.id"),split=" ")
mgi$high_MP_phen <- lapply(mgi$high_MP_ID,function(x) unique(mgi_allphens$phenotype[which(mgi_allphens$MP.id%in%x)]))

#allele info
mgi$allele_info <- lapply(mgi$MGI_ID, function(x) unique(mgi_df$allele_info[which(mgi_df$MGI_ID==x&mgi_df$is_lethal=="Y")]))

#is_lethal
mgi$is_lethal <- ifelse(lapply(mgi$lethal_MP_ID,length)==0,"N","Y")

# add on gene names and human ortholog names
mgi_names <- read.xlsx("Gene_lists/MGI_MPs/HMD_HumanPhenotype.xlsx")
mgi_names$MGI.Marker.Accession.ID <- gsub(" ", "", mgi_names$MGI.Marker.Accession.ID, fixed = TRUE)
mgi$mouse_symbol <- vlookup(mgi$MGI_ID,mgi_names,result_column="Mouse.Marker.Symbol",lookup_column = "MGI.Marker.Accession.ID")

names <- read.csv("Gene_lists/Universe/gene_with_protein_product.txt",sep = "\t", comment.char = "#",stringsAsFactors = FALSE)
mgi$human_symbol <- vlookup(mgi$MGI_ID,names,result_column="symbol",lookup_column = "mgd_id")

 
unmapped <- lapply(mgi$MGI_ID[which(!is.na(mgi$mouse_symbol)&is.na(mgi$human_symbol))], function(x) which(grepl(x,universe_df$mgd_id)==TRUE) )
lengths(unmapped)
remapped <- mgi$MGI_ID[which(!is.na(mgi$mouse_symbol)&is.na(mgi$human_symbol))]
mgi$human_symbol[which(mgi$MGI_ID%in%remapped[which(lengths(unmapped)==1)])] <- universe_df$gene[unlist(unmapped[which(lengths(unmapped)==1)])]
  
mgi$MGI_ID[which(!is.na(mgi$mouse_symbol)&is.na(mgi$human_symbol))]
mgi$human_symbol <- checkGeneSymbols(mgi$human_symbol,unmapped.as.na=FALSE)[[3]]


#remove non protein-coding genes
mgi <- mgi[-which(is.na(mgi$human_symbol)),]  

rm(allele,mgi_names,mgi_df,mgi_lethalphens,mgi_allphens,mgi_highphens,names,unmapped,remapped,lethalphens)
save(mgi, file="output/Data/mgi.rda", compress="bzip2")
write.xlsx(mgi,"output/spreadsheets/MGI_genes_with_phenotypes.xlsx")

```

#### IMPC
release 8.0 phenotype information downloaded from [ftp](ftp://ftp.ebi.ac.uk/pub/databases/impc/release-8.0/csv/)

* keep only homozygotes
* put together all MP terms and lethal MP terms
* add on human ortholog names
* remove non protein-coding genes


```{r impc, eval=FALSE}
impc_all<- read.csv("Gene_lists/IMPC/ALL_genotype_phenotype.csv",header=TRUE, sep = ",",fill=TRUE,stringsAsFactors = FALSE)
impc_all<-impc_all[-which(impc_all$mp_term_id==""|impc_all$zygosity!="homozygote"),]

lethalphens <- read.xlsx("Gene_lists/MGI_MPs/MGI_lethal_phenotypes.xlsx")
impc_all$is_lethal <- ifelse(impc_all$mp_term_id%in%lethalphens$MP.id,"Y","N")

impc <- data.frame(mgi_id=unique(impc_all$marker_accession_id))

#mp ids
impc$all_MP_ID <- lapply(impc$mgi_id, function(x) unique(impc_all$mp_term_id[which(impc_all$marker_accession_id==x)]))
impc$all_MP_phen <- lapply(impc$mgi_id, function(x) unique(impc_all$mp_term_name[which(impc_all$marker_accession_id==x)]))
impc$lethal_MP_ID <- lapply(impc$mgi_id, function(x) unique(impc_all$mp_term_id[which(impc_all$marker_accession_id==x&impc_all$is_lethal=="Y")]))
impc$lethal_MP_phen <- lapply(impc$mgi_id, function(x) unique(impc_all$mp_term_name[which(impc_all$marker_accession_id==x&impc_all$is_lethal=="Y")]))

impc$is_lethal <- rep("N",length(impc$mgi_id))
impc$is_lethal[which(lengths(impc$lethal_MP_ID)>0)]<-"Y"

# add on gene names and human ortholog names
names <- read.csv("Gene_lists/Universe/gene_with_protein_product.txt",sep = "\t", comment.char = "#",stringsAsFactors = FALSE)
impc$human_symbol <- vlookup(impc$mgi_id,names,result_column="symbol",lookup_column = "mgd_id")
impc$human_symbol <- checkGeneSymbols(impc$human_symbol,unmapped.as.na=FALSE)[[3]]
#removing pseudogenes/non-coding genes etc
impc <- impc[-which(is.na(impc$human_symbol)),]

rm(impc_all)
save(impc, file="output/Data/impc.rda", compress="bzip2")


```

### Cell KO data

Cell essentiality data downloaded from 3 papers
*FINISH LATER!!!*

```{r cell_ko, eval=FALSE}
#get wang, blomen, hart genes and essentialomes
#read wang genes and which are essential in different cell lines
wang <- read.xlsx("Gene_lists/Cell_KOs/wang_essentials.xlsx",startRow = 1, colNames = TRUE, rowNames = FALSE)
wang$Gene<- checkGeneSymbols(wang$Gene, unmapped.as.na=FALSE)[[3]]
wang<- wang[-which(duplicated(wang$Gene)=="TRUE"),] 

#read blomen genes and which are essential in different cell lines
blomen <- read.xlsx("Gene_lists/Cell_KOs/blomen_essentials.xlsx",startRow = 1, colNames = TRUE, rowNames = FALSE)
blomen$Gene<- checkGeneSymbols(blomen$Gene, unmapped.as.na=FALSE)[[3]]
blomen<- blomen[-which(duplicated(blomen$Gene)=="TRUE"),] 

#read hart genes and which are essential in different cell lines
hart <- read.xlsx("Gene_lists/Cell_KOs/hart_essentials.xlsx",startRow = 1, colNames = TRUE, rowNames = FALSE)
hart$Gene<- checkGeneSymbols(hart$Gene, unmapped.as.na=FALSE)[[3]]
hart<- hart[-which(duplicated(hart$Gene)=="TRUE"),] 

common_genes <- intersect(wang$Gene,intersect(hart$Gene,blomen$Gene))
#Seeing essentiality of each gene accross all cell lines in the three papers
cell_KOs <- data.frame(common_genes,ifelse(vlookup(common_genes,blomen,result_column = "hap1",lookup_column="Gene")=="essential",1,0),
                       ifelse(vlookup(common_genes,blomen,result_column = "kbm7",lookup_column="Gene")=="essential",1,0),ifelse(vlookup(common_genes,wang,result_column = "KBM7",lookup_column="Gene")=="essential",1,0),
                       ifelse(vlookup(common_genes,wang,result_column = "k562",lookup_column="Gene")=="essential",1,0),ifelse(vlookup(common_genes,wang,result_column = "jiyoye",lookup_column="Gene")=="essential",1,0),
                       ifelse(vlookup(common_genes,wang,result_column = "raji",lookup_column="Gene")=="essential",1,0),ifelse(vlookup(common_genes,hart,result_column = "hct",lookup_column="Gene")=="essential",1,0),
                       ifelse(vlookup(common_genes,hart,result_column = "hela",lookup_column="Gene")=="essential",1,0),ifelse(vlookup(common_genes,hart,result_column = "gbm",lookup_column="Gene")=="essential",1,0),
                       ifelse(vlookup(common_genes,hart,result_column = "rpe1",lookup_column="Gene")=="essential",1,0),ifelse(vlookup(common_genes,hart,result_column = "dld1",lookup_column="Gene")=="essential",1,0))
colnames(cell_KOs) <- c("Gene","blomen_hap1","blomen_kbm7","wang_kbm7","wang_k562","wang_jiyoye","wang_raji","hart_hct","hart_hela","hart_gbm","hart_rpe1","hart_dld1")
cell_KOs$no_hits <- cell_KOs$blomen_hap1+cell_KOs$blomen_kbm7+cell_KOs$wang_kbm7+cell_KOs$wang_k562+cell_KOs$wang_jiyoye+cell_KOs$wang_raji+
                        cell_KOs$hart_hct+cell_KOs$hart_hela+cell_KOs$hart_gbm+cell_KOs$hart_rpe1+cell_KOs$hart_dld1

save(cell_KOs, file="output/Data/cell_KOs.rda", compress="bzip2")
rm(blomen,hart,wang,common_genes)

```

### GO Annotations

### Human Lethal Genes- OMIM API Search


### Putting it all together
```{r universe, eval=FALSE}
#disease status, phenotype, inheritance pattern
universe_df$omim <- ifelse(universe_df$gene%in%omim$Gene,"Y",NA)
universe_df$mim_number <- vlookup(universe_df$gene,omim,result_column="Mim.Number",lookup_column="Gene")
universe_df$phenotype <- vlookup(universe_df$gene,omim,result_column="Phenotypes",lookup_column="Gene")
universe_df$Inheritance_pattern <- vlookup(universe_df$gene,omim,result_column="Inheritance_pattern",lookup_column="Gene")

rm(omim)
#exac scores
universe_df$exac <- ifelse(universe_df$gene%in%exac$gene,"Y",NA)
universe_df$mis_z <- vlookup(universe_df$gene,exac,result_column="mis_z",lookup_column="gene")
universe_df$syn_z <- vlookup(universe_df$gene,exac,result_column="syn_z",lookup_column="gene")
universe_df$pLI <- vlookup(universe_df$gene,exac,result_column="pLI",lookup_column="gene")
universe_df$constrained <- ifelse(universe_df$mis_z>=3.09|universe_df$pLI>=0.9,"Y","N")
rm(exac)

#regional constraint
universe_df$highest_ccr<-vlookup(universe_df$gene,highest_ccr,lookup_column = "gene",result_column = "ccr")
universe_df$ccr99<-ifelse(universe_df$gene%in%ccr99_genes$gene,"Y","N")
universe_df$ccr99_n<-vlookup(universe_df$gene,ccr99_genes,lookup_column = "gene",result_column = "n")
rm(ccr99_genes,highest_ccr)

universe_df$regional_missense_constraint <- ifelse(universe_df$gene%in%genes_w_var$gene,"Y","N")
universe_df$regional_missense_gamma<-vlookup(universe_df$gene,genes_w_var,lookup_column = "gene",result_column = "lowest_gamma")
rm(genes_w_var)

universe_df$nmd_min <-ifelse(universe_df$gene%in%nmd_min$gene,"Y","N")
universe_df$nmd_min_rank <- vlookup(universe_df$gene,nmd_min,lookup_column = "gene",result_column = "min.rank")
universe_df$nmd_min_rank<-universe_df$nmd_min_rank/0.01
rm(nmd_min)

universe_df$any_constraint <- ifelse(universe_df$mis_z>=3.09|universe_df$pLI>=0.9|universe_df$ccr99=="Y"|universe_df$regional_missense_constraint=="Y"|universe_df$nmd_min=="Y","Y","N")
universe_df$any_reg_constraint <- ifelse(universe_df$ccr99=="Y"|universe_df$regional_missense_constraint=="Y"|universe_df$nmd_min=="Y","Y","N")

#MGI mouse knockouts- lethal or non-lethal in a mouse
universe_df$MGI_ID <- vlookup(universe_df$gene,mgi,result_column="MGI_ID",lookup_column="human_symbol")
universe_df$lethal_MGI <- vlookup(universe_df$gene,mgi,result_column="is_lethal",lookup_column="human_symbol")
universe_df$lethal_MP_ID <- vlookup(universe_df$gene,mgi,result_column="lethal_MP_ID",lookup_column="human_symbol")
universe_df$lethal_MP_phen <- vlookup(universe_df$gene,mgi,result_column="lethal_MP_phen",lookup_column="human_symbol")
universe_df$all_MP_ID <- vlookup(universe_df$gene,mgi,result_column="all_MP_ID",lookup_column="human_symbol")
universe_df$all_MP_phen <- vlookup(universe_df$gene,mgi,result_column="all_MP_phen",lookup_column="human_symbol")
universe_df$high_MP_ID <- vlookup(universe_df$gene,mgi,result_column="high_MP_ID",lookup_column="human_symbol")
universe_df$high_MP_phen <- vlookup(universe_df$gene,mgi,result_column="high_MP_phen",lookup_column="human_symbol")
universe_df$allele_info <- vlookup(universe_df$gene,mgi,result_column="allele_info",lookup_column="human_symbol")
universe_df$mouse_symbol <- vlookup(universe_df$gene,mgi,result_column="mouse_symbol",lookup_column="human_symbol")

#IMPC phenotype data
universe_df$lethal_IMPC <- vlookup(universe_df$gene,impc,result_column="is_lethal",lookup_column="human_symbol")
universe_df$IMPC_all_MP_ID <- vlookup(universe_df$gene,impc,result_column="all_MP_ID",lookup_column="human_symbol")
universe_df$IMPC_all_MP_phen <- vlookup(universe_df$gene,impc,result_column="all_MP_phen",lookup_column="human_symbol")
universe_df$IMPC_lethal_MP_ID <- vlookup(universe_df$gene,impc,result_column="lethal_MP_ID",lookup_column="human_symbol")
universe_df$IMPC_lethal_MP_phen <- vlookup(universe_df$gene,impc,result_column="lethal_MP_phen",lookup_column="human_symbol")

universe_df$IMPC_ko <- rep(NA,length(universe_df$gene))
universe_df$IMPC_ko[which(lengths(universe_df$IMPC_all_MP_ID)>0)] <-  "Y"

universe_df$mouse_ko <- ifelse(!is.na(universe_df$MGI_ID),"Y",ifelse(universe_df$IMPC_ko=="Y","Y",NA))
universe_df$lethal_mouse <- rep(NA,length(universe_df$gene))
universe_df$lethal_mouse[which(universe_df$mouse_ko=="Y")] <- "N"
universe_df$lethal_mouse[which(universe_df$lethal_MGI=="Y"|universe_df$lethal_IMPC=="Y")] <- "Y"

rm(mgi,impc)


#cell knockouts
universe_df$cell_ko <- ifelse(universe_df$gene%in%cell_KOs$Gene,"Y",NA)
universe_df$cell_essential_hits <- vlookup(universe_df$gene,cell_KOs,result_column="no_hits",lookup_column="Gene")
universe_df$cell_essential <- ifelse(universe_df$cell_essential_hits>=3,"Y","N")
rm(cell_KOs)


#GO terms
universe_df$go_terms <- go_annotations$universe_df.go_terms
universe_df$go_names <- go_annotations$universe_df.go_names
rm(go_annotations)

#human lethal genes- are the genes in my OMIM API search for genes causing lethality in humans
load("output/Data/human_lethal_genes.rda")
universe_df$human_lethal_B <- ifelse(is.na(vlookup(universe_df$gene,lethal_genes[which(lethal_genes$listB=="yes"),],lookup_column = "gene")),"N","Y")
universe_df$human_lethal_A <- ifelse(is.na(vlookup(universe_df$gene,lethal_genes[which(lethal_genes$listA=="yes"),],lookup_column = "gene")),"N","Y")
universe_df$lethal_phen <- lapply(universe_df$gene,function(x) vlookup(x,lethal_genes[which(lethal_genes$listB=="yes"),],lookup_column = "gene",result_column = "lethal_phenotype_mim"))
universe_df$lethal_inheritance <- lapply(universe_df$gene,function(x) vlookup(x,lethal_genes[which(lethal_genes$listB=="yes"),],lookup_column = "gene",result_column = "lethal_inheritance_pattern"))
rm(lethal_genes)

save(universe_df, file="output/Data/universe_df.rda", compress="bzip2")
write.xlsx(universe_df[,-c(which(names(universe_df)=="go_terms"),which(names(universe_df)=="go_names"))],"output/spreadsheets/universe_all_info.xlsx",append=TRUE)


```


## Results/ Analysis

Seperated according to paper figures

### Functions for the plots 

 
```{r plot_fun}
blank_theme <- function() {
  theme_minimal(base_size=12,base_family="Helvetica") %+replace%
    theme(
      axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      axis.text = element_blank(),
      panel.border = element_blank(),
      panel.grid=element_blank(),
      axis.ticks = element_blank(),
      plot.title=element_text(size=12, face="bold"),
      legend.title=element_blank()
    )
}

scatter_theme <- function() {
  theme_bw(base_size=12,base_family="Helvetica") %+replace%
    theme(
      panel.grid=element_blank(),
      panel.border = element_blank(),
      axis.line = element_line(),
      plot.title=element_text(size=12, face="bold"),
      axis.title=element_text(size=12),
      axis.text=element_text(size=12)
    )
}

bar_theme <- function() {
  theme_bw(base_size=12,base_family="Helvetica") %+replace%
    theme(
      axis.title.x = element_blank(),
      panel.grid=element_blank(),
      panel.border = element_blank(),
      axis.ticks.x = element_blank(),
      axis.line = element_line(),
      legend.title=element_blank()
    )
}
```






